src/PostgREST/Error.hs

```diff
@@ -127,69 +108,151 @@ instance JSON.ToJSON H.QueryError where
-    "42883"   -> HT.status404 -- undefined function
-    "42P01"   -> HT.status404 -- undefined table
-    "42501"   -> if authed then HT.status403 else HT.status401 -- insufficient privilege
-    'P':'T':n -> fromMaybe HT.status500 (HT.mkStatus <$> readMaybe n <*> pure m)
-    _         -> HT.status400
-httpStatus _ (P.SessionError (H.QueryError _ _ (H.ResultError _))) = HT.status500
-httpStatus _ (P.SessionError (H.QueryError _ _ (H.ClientError _))) = HT.status503
+    "message" .= ("Database client error" :: Text),
+    "details" .= (fmap toS d :: Maybe Text)]
+
+pgErrorStatus :: Bool -> P.UsageError -> HT.Status
+pgErrorStatus _      (P.ConnectionError _)                                      = HT.status503
+pgErrorStatus _      (P.SessionError (H.QueryError _ _ (H.ClientError _)))      = HT.status503
+pgErrorStatus authed (P.SessionError (H.QueryError _ _ (H.ResultError rError))) =
+  case rError of
+    (H.ServerError c m _ _) ->
+      case toS c of
+        '0':'8':_ -> HT.status503 -- pg connection err
+        '0':'9':_ -> HT.status500 -- triggered action exception
+        '0':'L':_ -> HT.status403 -- invalid grantor
+        '0':'P':_ -> HT.status403 -- invalid role specification
+        "23503"   -> HT.status409 -- foreign_key_violation
+        "23505"   -> HT.status409 -- unique_violation
+        '2':'5':_ -> HT.status500 -- invalid tx state
+        '2':'8':_ -> HT.status403 -- invalid auth specification
+        '2':'D':_ -> HT.status500 -- invalid tx termination
+        '3':'8':_ -> HT.status500 -- external routine exception
+        '3':'9':_ -> HT.status500 -- external routine invocation
+        '3':'B':_ -> HT.status500 -- savepoint exception
+        '4':'0':_ -> HT.status500 -- tx rollback
+        '5':'3':_ -> HT.status503 -- insufficient resources
+        '5':'4':_ -> HT.status413 -- too complex
+        '5':'5':_ -> HT.status500 -- obj not on prereq state
+        '5':'7':_ -> HT.status500 -- operator intervention
+        '5':'8':_ -> HT.status500 -- system error
+        'F':'0':_ -> HT.status500 -- conf file error
+        'H':'V':_ -> HT.status500 -- foreign data wrapper error
+        "P0001"   -> HT.status400 -- default code for "raise"
+        'P':'0':_ -> HT.status500 -- PL/pgSQL Error
+        'X':'X':_ -> HT.status500 -- internal Error
+        "42883"   -> HT.status404 -- undefined function
+        "42P01"   -> HT.status404 -- undefined table
+        "42501"   -> if authed then HT.status403 else HT.status401 -- insufficient privilege
+        'P':'T':n -> fromMaybe HT.status500 (HT.mkStatus <$> readMaybe n <*> pure m)
+        _         -> HT.status400
+
+    _                       -> HT.status500
+
+checkIsFatal :: PgError -> Maybe Text
+checkIsFatal (PgError _ (P.ConnectionError e))
+  | isAuthFailureMessage = Just "Password authentication failed"
+  | otherwise = Nothing
+  where isAuthFailureMessage = "FATAL:  password authentication failed" `isPrefixOf` toS (fromMaybe "" e) 
```

Comment:
Instead of parsing the error response from PostgreSQL, could the error code somehow be obtained?

It should be `28P01` according to https://www.postgresql.org/docs/current/errcodes-appendix.html.

Reply:
I think we can't do it with the current interface of `Hasql.Pool` `use`. I think we would have to go a bit over hasql internals for this.

`use` returns an Either (as usual, left error, right data). The error given is an `UsageError`, which is itself a union type that encapsulates either `ConnectionError` or `QueryError`. 

There's also a bit of type rabbit-hole on `QueryError`, where eventually error codes may live, but the point is that bad password is a `ConnectionError`, which is simply defined as `type ConnectionError = Maybe ByteString`. The string doesn't contain any code.

Example with `stack ghci`:
```haskell
> import qualified Hasql.Pool as P
> P.acquire (1, 10, "postgres://goodUser:badPasswor@goodHost:goodPort/goodDbName") >>= (\pool -> P.use pool getPgVersion)

< Left (ConnectionError (Just "FATAL:  password authentication failed for user \"goodUser\"\nFATAL:  no pg_hba.conf entry for host \"goodHost\", user \"goodUser\", database \"goodDbName\", SSL off\n"))
```

Reply:
@Qu4tro True. I've just noticed that. Parsing the error should be good for now.

Reply:
Just opened an issue for exposing the error code in `ConnectionError`: https://github.com/nikita-volkov/hasql/issues/112.